### PR TITLE
feat: Kafka external streams support SCRAM-SHA-256/512

### LINKED ISSUE #453: Kafka External Stream: support SCRAM-SHA-256 and  SCRAM-SHA-512
**Describe what enhancement you'd like to have**

Currently, Kafka external streams only support one SASL mechnism `PLAIN`, in order to be able to connect to more Kafka deployments, we should also support SCRAM-SHA-256 and  SCRAM-SHA-512.

### PR DESCRIPTION
**Pull request by zliang-min**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:

closes #453 

Added support for SASL mechanisms SCRAM-SHA-256 and SCRAM-SHA-512.

### CODE REVIEW

File: src/Storages/ExternalStream/ExternalStreamSettings.h [hunk 1]
```diff
@@ -15,6 +15,7 @@ class ASTStorage;
     M(String, security_protocol, "plaintext", "The protocol to connection external logstore", 0) \
     M(String, username, "", "The username of external logstore", 0) \
     M(String, password, "", "The password of external logstore", 0) \
+    M(String, sasl_mechanism, "PLAIN", "SASL mechanism to use for authentication. Supported: PLAIN, SCRAM-SHA-256, SCRAM-SHA-512. Default to PLAIN when SASL is enabled.", 0) \
```
Comment (jovezhong) [inline src/Storages/ExternalStream/ExternalStreamSettings.h:18]: Do we support `GSSAPI`?

Comment (zliang-min) [reply src/Storages/ExternalStream/ExternalStreamSettings.h:18]: If you have such a broker, I can run some tests.

File: src/KafkaLog/KafkaWALSettings.h [hunk 2]
```diff
@@ -23,6 +22,31 @@ struct KafkaWALAuth
...
+    }
+
+    bool usesSecureConnection() const
+    {
+        return boost::iends_with(security_protocol, "SSL");
+    }
+
+    void populateConfigs(std::vector<std::pair<std::string, std::string>> & params) const
+    {
+        params.emplace_back("security.protocol", security_protocol.c_str());
```
Comment (chenziliang) [inline src/KafkaLog/KafkaWALSettings.h]: do we need explicitly call `c_str()` ?

File: src/KafkaLog/KafkaWALSettings.h [hunk 3]
```diff
@@ -38,14 +62,15 @@ struct KafkaWALSettings
         .security_protocol = "plaintext",
         .username = "",
         .password = "",
+        .sasl_mechanism = "",
```
Comment (chenziliang) [inline src/KafkaLog/KafkaWALSettings.h]: Look like this explicit empty init is not necessary ?

File: src/Storages/ExternalStream/Kafka/Kafka.cpp [hunk 4]
```diff
@@ -144,6 +144,23 @@ void Kafka::cacheVirtualColumnNamesAndTypes()
...
+            .password = settings->password.value,
+            .sasl_mechanism = settings->sasl_mechanism.value,
+            .ssl_ca_cert_file = settings->ssl_ca_cert_file.value,
+        };
+}
+
+klog::KafkaWALSimpleConsumerPtr Kafka::getConsumer(int32_t fetch_wait_max_ms) const
+{
+
+        return klog::KafkaWALPool::instance(nullptr).getOrCreateStreamingExternal(settings->brokers.value, auth(), fetch_wait_max_ms);
```
Comment (chenziliang) [inline src/Storages/ExternalStream/Kafka/Kafka.cpp]: It may not matter, but we have a bunch of memory copy every time we call getConsumer for auth

**Review by chenziliang — COMMENTED** [review]
In general, looks good.

File: src/KafkaLog/KafkaWALSettings.h (hunk 3 above)
Comment (zliang-min) [reply src/KafkaLog/KafkaWALSettings.h]: I was following the existing code. I will remove all ` = ""` and just keep `.security_protocol = "plaintext'`.

File: src/KafkaLog/KafkaWALSettings.h (hunk 2 above)
Comment (zliang-min) [reply src/KafkaLog/KafkaWALSettings.h]: `c_str()` is not needed. I was just moving the existing code around to refactor it, thus I didn't touch it. I removed all `c_str()` calls in the update.

File: src/Storages/ExternalStream/Kafka/Kafka.cpp (hunk 4 above)
Comment (zliang-min) [reply src/Storages/ExternalStream/Kafka/Kafka.cpp]: The auth struct is small and this is not in a code hot path, copying looks fine to me. But I changed `auth()` just returns a reference instead anyways.

**Review by zliang-min — COMMENTED** [review]
@chenziliang PR updated per your comments, please take another, thanks!

**Review by chenziliang — APPROVED** [review]

**Comment by jovezhong** [comment]
(Jove Github Bot) added it to the current sprint.

**Comment by Neustradamus** [comment]
@zliang-min: Nice!

Linked to:
- https://github.com/scram-sasl/info/issues/1

**Comment by jovezhong** [comment]
Hi @Neustradamus , thanks for linking this ticket with the collective list about scram-sasl. Timeplus Cloud supports scram-sasl long time ago (via benthos) and this week we added the native support in Proton. So with a single binary, data engineers can connect to their Kafka clusters in their preferred security model.

**Comment by jovezhong** [comment]
(Jove Github Bot) moved this ticket out of the GitHub project(up to 1200 tickets for one project).

**Pull request merged**